File: frontend/src/lib/api/timeline.ts
import { apiFetch } from './client'
import type { ApiSlot } from './types'

export type ApiEventKind = 'diet' | 'prep' | 'med' | 'meal' | 'fast' | 'arrive' | 'stool'

export type ApiTimelineEvent = {
  id: string
  at: string
  kind: ApiEventKind
  title: string
  detail: string
  tentative: boolean
  agent: string | null
  prep_image_label: string | null
  sort_order: number
}

export type ApiTimeline = {
  public_code: string
  protocol_name: string
  source_label: string
  version_id: number
  version_label: string
  procedure_date: string
  slot: ApiSlot
  reporting_time: string
  events: ApiTimelineEvent[]
}

export function getApiTimeline(publicCode: string) {
  return apiFetch<ApiTimeline>(
    `/api/sessions/${encodeURIComponent(publicCode)}/timeline`,
  )
}

File: backend/app/schemas/timeline.py
from __future__ import annotations

from datetime import date, datetime, time
from typing import Literal, Optional

from pydantic import BaseModel, Field


EventKind = Literal["diet", "prep", "med", "meal", "fast", "arrive", "stool"]


class TimelineEventOut(BaseModel):
    id: str
    at: datetime
    kind: EventKind
    title: str
    detail: str = ""
    tentative: bool = False
    agent: Optional[str] = None
    prep_image_label: Optional[str] = None
    sort_order: int = 0


class TimelineOut(BaseModel):
    public_code: str
    protocol_name: str
    source_label: str = ""
    version_id: int
    version_label: str
    procedure_date: date
    slot: str
    reporting_time: time
    events: list[TimelineEventOut] = Field(default_factory=list)
